### DISCUSSION TITLE
Add own risk measure

### DISCUSSION THREAD

**Discussion by Conway-Yu** [skfolio/skfolio#75] [Q&A]
Hello,

Thank you for the amazing package.

After working with it for a little while, I want to add other risk measures for optimization.
To be specific, I was looking at the duration spread ratio optimization defined as:

$DSR(w_t) = \frac{w_t E_t}{{\sqrt{D^2_{t}+ S^2_{t}}}}$

where $D$ is duration and $S$ is spread. Thus, one needs to add $\sqrt{D^2_{t}+ S^2_{t}}$ as a risk measure.

However, I do not really get it to work.

Does someone have some guidance?

Thank you very much.

**Comment by HugoDelatte** [comment]
Hello, thank you!
The risk  $\sqrt{D^2_{t}+ S^2_{t}}$  should be a function of the weights vector $w$. 
I couldn't find literature about this risk, only about Duration Times Spread (DTS). If you have a link that would be useful.

**Reply by Conway-Yu** [reply]
Hello, thanks for the reply.

You can find information about the risk measure in the following link in equation 14:
https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4825814

$D_t$ is indeed $w_t' D_t$, and $S_t$ is $w_t' S_t$.

This risk metric in the end did not find a huge usecase in my specific optimization problem. Thus, I did not think of a nice way to implement it into skfolio yet. If I have more time at a later point in time, I am thinking of doing it though.

However, this (quite) dirty approach might already help someone:

Add `assets_duration` and `assets_spread` as attributes and include the following code into the `ConvexOptimization` Class:
```
    def _duration_spread_risk(self, prior_model: PriorModel, w: cp.Variable) -> skt.RiskResult:
        portfolio_duration = w @ self.assets_duration.iloc[-1, :]
        portfolio_spread = w @ self.assets_spread.iloc[-1, :]

        # DSR: sqrt((D'w)^2 + (S'w)^2)
        risk = cp.norm(cp.hstack([portfolio_duration, portfolio_spread]), 2)

        constraints = []
        return risk, constraints
```
Also modify the `fit` function in the `MeanRisk` Class after all the risk metric analysis, before matching the objective function
```
        if self.assets_duration is not None and self.assets_spread is not None:
            risk, _ = self._duration_spread_risk(prior_model=prior_model, w=w)
```
Please be aware that I also implemented my own `walkforward` function and set the attributes `assets_duration` and `assets_spread` for the estimator in each walkforward step, so I always guarantee to take the correct duration and spread via `iloc[-1,:]`.

**Reply by HugoDelatte** [reply]
Very interesting, thanks for sharing it.
Your implementation is the right approach. 
I believe `assets_duration` and  `assets_spread` are time series? If so, we will implement it using the metadata routing API from scikit-learn that has been merged recently and move them to the `prior_model`.
By doing that, WalkForward will naturally work.
